<!DOCTYPE html>

<html :class="{'dark': darkMode === 'dark' || (darkMode === 'system' &amp;&amp; window.matchMedia('(prefers-color-scheme: dark)').matches)}" class="scroll-smooth" data-content_root="./" lang="en" x-data="{ darkMode: localStorage.getItem('darkMode') || localStorage.setItem('darkMode', 'system'), activeSection: '' }" x-init="$watch('darkMode', val =&gt; localStorage.setItem('darkMode', val))">
<head>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<meta charset="utf-8"/>
<meta content="white" media="(prefers-color-scheme: light)" name="theme-color"/>
<meta content="black" media="(prefers-color-scheme: dark)" name="theme-color"/>
<meta content="width=device-width, initial-scale=1" name="viewport"/>
<title>Quick Start | H2M 1.0.0 documentation</title>
<meta content="Quick Start | H2M 1.0.0 documentation" property="og:title"/>
<meta content="Quick Start | H2M 1.0.0 documentation" name="twitter:title"/>
<link href="_static/pygments.css?v=8d216cef" rel="stylesheet" type="text/css"/>
<link href="_static/theme.css?v=8958d2b4" rel="stylesheet" type="text/css"/>
<link href="_static/css/custom.css" rel="stylesheet" type="text/css"/>
<link href="search.html" rel="search" title="Search"/>
<link href="genindex.html" rel="index" title="Index"/>
<link href="documentation.html" rel="next" title="Complete Documentation"/>
<link href="index.html" rel="prev" title="H2M Documentation"/>
<script>
    <!-- Prevent Flash of wrong theme -->
      const userPreference = localStorage.getItem('darkMode');
      let mode;
      if (userPreference === 'dark' || window.matchMedia('(prefers-color-scheme: dark)').matches) {
        mode = 'dark';
        document.documentElement.classList.add('dark');
      } else {
        mode = 'light';
      }
      if (!userPreference) {localStorage.setItem('darkMode', mode)}
    </script>
</head>
<body :class="{ 'overflow-hidden': showSidebar }" class="min-h-screen font-sans antialiased bg-background text-foreground" x-data="{ showSidebar: false }">
<div @click.self="showSidebar = false" class="fixed inset-0 z-50 overflow-hidden bg-background/80 backdrop-blur-sm md:hidden" x-cloak="" x-show="showSidebar"></div><div class="relative flex flex-col min-h-screen" id="page"><a class="absolute top-0 left-0 z-[100] block bg-background p-4 text-xl transition -translate-x-full opacity-0 focus:translate-x-0 focus:opacity-100" href="#content">
      Skip to content
    </a><header class="sticky top-0 z-40 w-full border-b shadow-sm border-border supports-backdrop-blur:bg-background/60 bg-background/95 backdrop-blur"><div class="container flex items-center h-14">
<div class="hidden mr-4 md:flex">
<a class="flex items-center mr-6" href="index.html">
<img alt="Logo" class="mr-2 hidden dark:block" height="24" src="_static/h2m-logo-final.png" width="24"/>
<img alt="Logo" class="mr-2 dark:hidden" height="24" src="_static/h2m-logo-final.png" width="24"/><span class="hidden font-bold sm:inline-block text-clip whitespace-nowrap">H2M 1.0.0 documentation</span>
</a></div><button @click="showSidebar = true" class="inline-flex items-center justify-center h-10 px-0 py-2 mr-2 text-base font-medium transition-colors rounded-md hover:text-accent-foreground hover:bg-transparent md:hidden" type="button">
<svg aria-hidden="true" fill="currentColor" height="24" viewbox="0 96 960 960" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M152.587 825.087q-19.152 0-32.326-13.174t-13.174-32.326q0-19.152 13.174-32.326t32.326-13.174h440q19.152 0 32.326 13.174t13.174 32.326q0 19.152-13.174 32.326t-32.326 13.174h-440Zm0-203.587q-19.152 0-32.326-13.174T107.087 576q0-19.152 13.174-32.326t32.326-13.174h320q19.152 0 32.326 13.174T518.087 576q0 19.152-13.174 32.326T472.587 621.5h-320Zm0-203.587q-19.152 0-32.326-13.174t-13.174-32.326q0-19.152 13.174-32.326t32.326-13.174h440q19.152 0 32.326 13.174t13.174 32.326q0 19.152-13.174 32.326t-32.326 13.174h-440ZM708.913 576l112.174 112.174q12.674 12.674 12.674 31.826t-12.674 31.826Q808.413 764.5 789.261 764.5t-31.826-12.674l-144-144Q600 594.391 600 576t13.435-31.826l144-144q12.674-12.674 31.826-12.674t31.826 12.674q12.674 12.674 12.674 31.826t-12.674 31.826L708.913 576Z"></path>
</svg>
<span class="sr-only">Toggle navigation menu</span>
</button>
<div class="flex items-center justify-between flex-1 space-x-2 sm:space-x-4 md:justify-end">
<div class="flex-1 w-full md:w-auto md:flex-none"><form @keydown.k.window.meta="$refs.search.focus()" action="search.html" class="relative flex items-center group" id="searchbox" method="get">
<input aria-label="Search the docs" class="inline-flex items-center font-medium transition-colors bg-transparent focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 ring-offset-background border border-input hover:bg-accent focus:bg-accent hover:text-accent-foreground focus:text-accent-foreground hover:placeholder-accent-foreground py-2 px-4 relative h-9 w-full justify-start rounded-[0.5rem] text-sm text-muted-foreground sm:pr-12 md:w-40 lg:w-64" id="search-input" name="q" placeholder="Search ..." type="search" x-ref="search"/>
<kbd class="pointer-events-none absolute right-1.5 top-2 hidden h-5 select-none text-muted-foreground items-center gap-1 rounded border border-border bg-muted px-1.5 font-mono text-[10px] font-medium opacity-100 sm:flex group-hover:bg-accent group-hover:text-accent-foreground">
<span class="text-xs">⌘</span>
    K
  </kbd>
</form>
</div>
<nav class="flex items-center space-x-1">
<button @click="darkMode = darkMode === 'light' ? 'dark' : 'light'" aria-label="Color theme switcher" class="relative inline-flex items-center justify-center px-0 text-sm font-medium transition-colors rounded-md hover:bg-accent hover:text-accent-foreground h-9 w-9" type="button">
<svg class="absolute transition-all scale-100 rotate-0 dark:-rotate-90 dark:scale-0" fill="currentColor" height="24" viewbox="0 96 960 960" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M480 685q45.456 0 77.228-31.772Q589 621.456 589 576q0-45.456-31.772-77.228Q525.456 467 480 467q-45.456 0-77.228 31.772Q371 530.544 371 576q0 45.456 31.772 77.228Q434.544 685 480 685Zm0 91q-83 0-141.5-58.5T280 576q0-83 58.5-141.5T480 376q83 0 141.5 58.5T680 576q0 83-58.5 141.5T480 776ZM80 621.5q-19.152 0-32.326-13.174T34.5 576q0-19.152 13.174-32.326T80 530.5h80q19.152 0 32.326 13.174T205.5 576q0 19.152-13.174 32.326T160 621.5H80Zm720 0q-19.152 0-32.326-13.174T754.5 576q0-19.152 13.174-32.326T800 530.5h80q19.152 0 32.326 13.174T925.5 576q0 19.152-13.174 32.326T880 621.5h-80Zm-320-320q-19.152 0-32.326-13.174T434.5 256v-80q0-19.152 13.174-32.326T480 130.5q19.152 0 32.326 13.174T525.5 176v80q0 19.152-13.174 32.326T480 301.5Zm0 720q-19.152 0-32.326-13.17Q434.5 995.152 434.5 976v-80q0-19.152 13.174-32.326T480 850.5q19.152 0 32.326 13.174T525.5 896v80q0 19.152-13.174 32.33-13.174 13.17-32.326 13.17ZM222.174 382.065l-43-42Q165.5 327.391 166 308.239t13.174-33.065q13.435-13.674 32.587-13.674t32.065 13.674l42.239 43q12.674 13.435 12.555 31.706-.12 18.272-12.555 31.946-12.674 13.674-31.445 13.413-18.772-.261-32.446-13.174Zm494 494.761-42.239-43q-12.674-13.435-12.674-32.087t12.674-31.565Q686.609 756.5 705.38 757q18.772.5 32.446 13.174l43 41.761Q794.5 824.609 794 843.761t-13.174 33.065Q767.391 890.5 748.239 890.5t-32.065-13.674Zm-42-494.761Q660.5 369.391 661 350.62q.5-18.772 13.174-32.446l41.761-43Q728.609 261.5 747.761 262t33.065 13.174q13.674 13.435 13.674 32.587t-13.674 32.065l-43 42.239q-13.435 12.674-31.706 12.555-18.272-.12-31.946-12.555Zm-495 494.761Q165.5 863.391 165.5 844.239t13.674-32.065l43-42.239q13.435-12.674 32.087-12.674t31.565 12.674Q299.5 782.609 299 801.38q-.5 18.772-13.174 32.446l-41.761 43Q231.391 890.5 212.239 890t-33.065-13.174ZM480 576Z"></path>
</svg>
<svg class="absolute transition-all scale-0 rotate-90 dark:rotate-0 dark:scale-100" fill="currentColor" height="24" viewbox="0 96 960 960" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M480 936q-151 0-255.5-104.5T120 576q0-138 90-239.5T440 218q25-3 39 18t-1 44q-17 26-25.5 55t-8.5 61q0 90 63 153t153 63q31 0 61.5-9t54.5-25q21-14 43-1.5t19 39.5q-14 138-117.5 229T480 936Zm0-80q88 0 158-48.5T740 681q-20 5-40 8t-40 3q-123 0-209.5-86.5T364 396q0-20 3-40t8-40q-78 32-126.5 102T200 576q0 116 82 198t198 82Zm-10-270Z"></path>
</svg>
</button>
</nav>
</div>
</div>
</header>
<div class="flex-1"><div class="container flex-1 items-start md:grid md:grid-cols-[220px_minmax(0,1fr)] md:gap-6 lg:grid-cols-[240px_minmax(0,1fr)] lg:gap-10"><aside :aria-hidden="!showSidebar" :class="{ 'translate-x-0': showSidebar }" class="fixed inset-y-0 left-0 md:top-14 z-50 md:z-30 bg-background md:bg-transparent transition-all duration-100 -translate-x-full md:translate-x-0 ml-0 p-6 md:p-0 md:-ml-2 md:h-[calc(100vh-3.5rem)] w-5/6 md:w-full shrink-0 overflow-y-auto border-r border-border md:sticky" id="left-sidebar">
<a class="!justify-start text-sm md:!hidden bg-background" href="index.html">
<img alt="Logo" class="mr-2 hidden dark:block" height="16" src="_static/h2m-logo-final.png" width="16"/>
<img alt="Logo" class="mr-2 dark:hidden" height="16" src="_static/h2m-logo-final.png" width="16"/><span class="font-bold text-clip whitespace-nowrap">H2M 1.0.0 documentation</span>
</a>
<div class="relative overflow-hidden md:overflow-auto my-4 md:my-0 h-[calc(100vh-8rem)] md:h-auto">
<div class="overflow-y-auto h-full w-full relative pr-6"><nav class="table w-full min-w-full my-6 lg:my-8">
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="documentation.html">Complete Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="jupyter.html">Jupyter Notebook Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream.html">COMING SOON - H2M Database Online Portal</a></li>
<li class="toctree-l1"><a class="reference internal" href="about.html">About H2M</a></li>
</ul>
</nav>
</div>
</div>
<button @click="showSidebar = false" class="absolute md:hidden right-4 top-4 rounded-sm opacity-70 transition-opacity hover:opacity-100" type="button">
<svg class="h-4 w-4" fill="currentColor" height="24" stroke="none" viewbox="0 96 960 960" width="24" xmlns="http://www.w3.org/2000/svg">
<path d="M480 632 284 828q-11 11-28 11t-28-11q-11-11-11-28t11-28l196-196-196-196q-11-11-11-28t11-28q11-11 28-11t28 11l196 196 196-196q11-11 28-11t28 11q11 11 11 28t-11 28L536 576l196 196q11 11 11 28t-11 28q-11 11-28 11t-28-11L480 632Z"></path>
</svg>
</button>
</aside>
<main class="relative py-6 lg:gap-10 lg:py-8 xl:grid xl:grid-cols-[1fr_300px]">
<div class="w-full min-w-0 mx-auto">
<nav aria-label="breadcrumbs" class="flex items-center mb-4 space-x-1 text-sm text-muted-foreground">
<a class="overflow-hidden text-ellipsis whitespace-nowrap hover:text-foreground" href="index.html">
<span class="hidden md:inline">H2M 1.0.0 documentation</span>
<svg aria-label="Home" class="md:hidden" fill="currentColor" height="18" stroke="none" viewbox="0 96 960 960" width="18" xmlns="http://www.w3.org/2000/svg">
<path d="M240 856h120V616h240v240h120V496L480 316 240 496v360Zm-80 80V456l320-240 320 240v480H520V696h-80v240H160Zm320-350Z"></path>
</svg>
</a>
<div class="mr-1">/</div><span aria-current="page" class="font-medium text-foreground overflow-hidden text-ellipsis whitespace-nowrap">Quick Start</span>
</nav>
<div id="content" role="main">
<section id="quick-start">
<h1>Quick Start<a class="headerlink" href="#quick-start" title="Link to this heading"><span>#</span></a></h1>
<a class="reference internal image-reference" href="_images/h2m-logo-final.png"><img alt="_images/h2m-logo-final.png" class="align-left" src="_images/h2m-logo-final.png" style="width: 150px;"/></a>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Data used in this tutorial can be downloaded from <a href="https://www.dropbox.com/scl/fo/1wtrnc9w6s9gemweuw2fv/h?rlkey=hli1z6tv096cjwit5oi6bwggg&amp;dl=0" rel="noopener noreferrer" target="_blank">this Dropbox folder</a>.</p>
</div>
<section id="package-installation">
<h2>Package Installation<a class="headerlink" href="#package-installation" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#package-installation'"><span>#</span></a></h2>
<p>H2M is available through the python package index (PyPI). To install, use pip:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">pip</span> <span class="n">install</span> <span class="n">bioh2m</span>
</pre></div>
</div>
<div class="admonition attention">
<p class="admonition-title">Attention</p>
<p>Python <strong>3.9-3.12</strong> are recommended since H2M has been tested compatible in them.</p>
</div>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>H2M has <code class="docutils literal notranslate"><span class="pre">pysam</span></code> as a dependency. This is for a function that can read .vcf files. If you are experiencing installation problems due to pysam, you can download and install the wheel file in <a class="reference external" href="https://github.com/kexindon/h2m-public/tree/main/install-wheels">the GitHub repository</a> without this function and the pysam dependency, which has been tested to solve most installation issues. The function rounded off in mini-h2m is also given in the repository.</p>
</div>
</section>
<section id="importing-packages">
<h2>Importing packages<a class="headerlink" href="#importing-packages" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#importing-packages'"><span>#</span></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="kn">import</span> <span class="nn">bioh2m</span> <span class="k">as</span> <span class="nn">h2m</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</pre></div>
</div>
</section>
<section id="loading-data">
<h2>Loading data<a class="headerlink" href="#loading-data" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#loading-data'"><span>#</span></a></h2>
<p>We should upload reference genome and GENCODE annotation data for both human and mouse, which could be directly downloaded from <a href="https://www.dropbox.com/scl/fo/1wtrnc9w6s9gemweuw2fv/h?rlkey=hli1z6tv096cjwit5oi6bwggg&amp;dl=0" rel="noopener noreferrer" target="_blank">this Dropbox folder</a>.<br/>
Both GRCh37 and GRCh38 human reference genome assemblys are available. Upload the one that you are going to use.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">path_h_ref</span><span class="p">,</span> <span class="n">path_m_ref</span> <span class="o">=</span> <span class="s1">'.../GCF_000001405.25_GRCh37.p13_genomic.fna.gz'</span><span class="p">,</span> <span class="s1">'.../GCF_000001635.27_GRCm39_genomic.fna.gz'</span>
    <span class="c1"># remember to replace the paths with yours; for human, GRCh38 reference genome assembly is also provided  </span>
    <span class="n">records_h</span><span class="p">,</span> <span class="n">index_list_h</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">genome_loader</span><span class="p">(</span><span class="n">path_h_ref</span><span class="p">)</span>
    <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span>  <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">genome_loader</span><span class="p">(</span><span class="n">path_m_ref</span><span class="p">)</span>

    <span class="n">path_h_anno</span><span class="p">,</span> <span class="n">path_m_anno</span> <span class="o">=</span> <span class="s1">'.../gencode_v19_GRCh37.db'</span><span class="p">,</span> <span class="s1">'.../gencode_vm33_GRCm39.db'</span>
    <span class="c1"># remember to replace the paths with yours</span>
    <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">anno_loader</span><span class="p">(</span><span class="n">path_h_anno</span><span class="p">),</span> <span class="n">h2m</span><span class="o">.</span><span class="n">anno_loader</span><span class="p">(</span><span class="n">path_m_anno</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="batch-processing">
<h2>Batch Processing<a class="headerlink" href="#batch-processing" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#batch-processing'"><span>#</span></a></h2>
<section id="input-format">
<h3>Input format<a class="headerlink" href="#input-format" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#input-format'"><span>#</span></a></h3>
<p>Common mutation data formats include <strong>MAF</strong> (Mutation Annotation Format, used by cBioPortal), <strong>VCF</strong> (Variant Call Format, used by genomAD), and <strong>ClinVar</strong> (a modified VCF format, used by ClinVar). Mutation coordinates, reference and alternative sequences are recorded in slightly different ways between the three.</p>
<a class="reference internal image-reference" href="_images/format.png"><img alt="_images/format.png" class="align-center" src="_images/format.png" style="width: 600px;"/></a>
<p>In batch processing, <strong>H2M accepts MAF input</strong>. More information about MAF format can be found at <a href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/#:~:text=Mutation" rel="noopener noreferrer" target="_blank">GDC Documentation</a>.</p>
<p>For MAF files, you need to build a pandas dataframe with columns as the following example:</p>
<a class="reference internal image-reference" href="_images/1.png"><img alt="_images/1.png" class="align-center" src="_images/1.png" style="width: 600px;"/></a>
<p>For VCF and ClinVar files, you will need to convert the mutation coordinates and sequence information to MAF format after this.</p>
<p>This can be achieved simply by using H2M built-in functions.</p>
<section id="read-from-cbioportal-maf">
<h4>Read from cBioPortal - MAF<a class="headerlink" href="#read-from-cbioportal-maf" title="Link to this heading"><span>#</span></a></h4>
<p>This format is compatible with all of the datasets in the <a href="https://www.cbioportal.org" rel="noopener noreferrer" target="_blank">cBioPortal</a>, as well as <a href="https://www.cancer.gov/ccg/research/genome-sequencing/tcga" rel="noopener noreferrer" target="_blank">TCGA</a> and <a href="https://www.synapse.org/#!Synapse:syn7222066/wiki/405659" rel="noopener noreferrer" target="_blank">AACR-GENIE</a>. Download the txt mutation data file from such public dataset and then load it as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">path_aacr</span> <span class="o">=</span> <span class="s1">'.../data_mutations_extended.txt'</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">cbio_reader</span><span class="p">(</span><span class="n">path_aacr</span><span class="p">)</span>
    <span class="n">df</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/cbio_reader.png"><img alt="_images/cbio_reader.png" class="align-center" src="_images/cbio_reader.png" style="width: 600px;"/></a>
</section>
<section id="read-from-genomad-vcf">
<h4>Read from GenomAD  - VCF<a class="headerlink" href="#read-from-genomad-vcf" title="Link to this heading"><span>#</span></a></h4>
<p>Search a specific gene in <a href="https://gnomad.broadinstitute.org" rel="noopener noreferrer" target="_blank">GenomAD browser</a>, and download the conluson csv.</p>
<a class="reference internal image-reference" href="_images/genomad.png"><img alt="_images/genomad.png" class="align-center" src="_images/genomad.png" style="width: 600px;"/></a>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="c1"># downloaded TP53 variants from genomAD</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">vcf_reader</span><span class="p">(</span><span class="s1">'.../gnomAD_v4.1.0_ENSG00000141510.csv'</span><span class="p">,</span><span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">'gene_name_h'</span><span class="p">]</span> <span class="o">=</span> <span class="s1">'TP53'</span>
    <span class="n">df</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/vcf_reader.png"><img alt="_images/vcf_reader.png" class="align-center" src="_images/vcf_reader.png" style="width: 600px;"/></a>
<p>Then convert it to <strong>MAF</strong> format.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">df</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">vcf_to_maf</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="n">df</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/vcf_to_maf.png"><img alt="_images/vcf_to_maf.png" class="align-center" src="_images/vcf_to_maf.png" style="width: 600px;"/></a>
</section>
<section id="read-from-clinvar">
<h4>Read from ClinVar<a class="headerlink" href="#read-from-clinvar" title="Link to this heading"><span>#</span></a></h4>
<p>download a <a href="https://www.ncbi.nlm.nih.gov/clinvar/" rel="noopener noreferrer" target="_blank">ClinVar</a> vcf.gz file, and choose your desired Variation IDs that you wish to model. These vcf.gz files are available in <a href="https://www.dropbox.com/scl/fo/1wtrnc9w6s9gemweuw2fv/h?rlkey=hli1z6tv096cjwit5oi6bwggg&amp;dl=0" rel="noopener noreferrer" target="_blank">this Dropbox folder</a>.</p>
<a class="reference internal image-reference" href="_images/clinvar.png"><img alt="_images/clinvar.png" class="align-center" src="_images/clinvar.png" style="width: 600px;"/></a>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">filepath</span> <span class="o">=</span> <span class="s1">'/Users/kexindong/Documents/GitHub/Database/PublicDatabase/ClinVar/GRCh37_clinvar_20240206.vcf.gz'</span>
    <span class="n">variation_ids</span> <span class="o">=</span> <span class="p">[</span><span class="mi">32798013</span><span class="p">,</span> <span class="mi">375926</span><span class="p">,</span> <span class="mi">325626</span><span class="p">,</span> <span class="mi">140953</span><span class="p">,</span> <span class="mi">233866</span><span class="p">,</span> <span class="mi">1796995</span><span class="p">,</span> <span class="mi">17578</span><span class="p">,</span> <span class="mi">573320</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">clinvar_reader</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">variation_ids</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">clinvar_to_maf</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">'gene_name_h'</span><span class="p">,</span>	<span class="s1">'start_h'</span><span class="p">,</span><span class="s1">'end_h'</span><span class="p">,</span><span class="s1">'ref_seq_h'</span><span class="p">,</span><span class="s1">'alt_seq_h'</span><span class="p">,</span><span class="s1">'type_h'</span><span class="p">,</span><span class="s1">'format'</span><span class="p">,</span><span class="s1">'ID'</span><span class="p">]]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">'ID'</span><span class="p">:</span><span class="s1">'index'</span><span class="p">})</span>
    <span class="n">df</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/clinvar_result.png"><img alt="_images/clinvar_result.png" class="align-center" src="_images/clinvar_result.png" style="width: 600px;"/></a>
</section>
</section>
<section id="get-canonical-transcript-ids-for-human">
<h3>Get canonical transcript IDs for human<a class="headerlink" href="#get-canonical-transcript-ids-for-human" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#get-canonical-transcript-ids-for-human'"><span>#</span></a></h3>
<p>There will be returning two dataframes for success and failures.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">df</span><span class="p">,</span> <span class="n">df_fail</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_batch</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s1">'h'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
    <span class="n">df</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/2.png"><img alt="_images/2.png" class="align-center" src="_images/2.png" style="width: 600px;"/></a>
</section>
<section id="query-orthologous-genes">
<h3>Query orthologous genes<a class="headerlink" href="#query-orthologous-genes" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#query-orthologous-genes'"><span>#</span></a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">df_queried</span><span class="p">,</span> <span class="n">df_fail</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">query_batch</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s1">'h2m'</span><span class="p">)</span>
    <span class="n">df_queried</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/3.png"><img alt="_images/3.png" class="align-center" src="_images/3.png" style="width: 700px;"/></a>
</section>
<section id="get-canonical-transcript-ids-for-mouse">
<h3>Get canonical transcript IDs for mouse<a class="headerlink" href="#get-canonical-transcript-ids-for-mouse" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#get-canonical-transcript-ids-for-mouse'"><span>#</span></a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">df_queried</span><span class="p">,</span> <span class="n">df_fail</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_batch</span><span class="p">(</span><span class="n">df_queried</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s1">'m'</span><span class="p">)</span>
    <span class="n">df_queried</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/4.png"><img alt="_images/4.png" class="align-center" src="_images/4.png" style="width: 700px;"/></a>
</section>
<section id="compute-the-muerine-variant-equivalents">
<h3>Compute the muerine variant equivalents<a class="headerlink" href="#compute-the-muerine-variant-equivalents" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#compute-the-muerine-variant-equivalents'"><span>#</span></a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">df_result</span><span class="p">,</span> <span class="n">df_fail</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model_batch</span><span class="p">(</span><span class="n">df_queried</span><span class="p">,</span> <span class="n">records_h</span><span class="p">,</span> <span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="mi">37</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="single-variant-input">
<h2>Single variant input<a class="headerlink" href="#single-variant-input" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#single-variant-input'"><span>#</span></a></h2>
<section id="id1">
<h3>Query orthologous genes<a class="headerlink" href="#id1" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#id1'"><span>#</span></a></h3>
<p>First of all, you can use H2M to query a human gene for the presence of mouse homologs and vice versa.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">query_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">'TP53'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    Query human gene: TP53;
    Mouse ortholog(s): Trp53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">query_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">'Trp53'</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s1">'m2h'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    Query human gene: Trp53;
    Mouse ortholog(s): TP53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.
</pre></div>
</div>
<p>The output is a list of information for all the mouse ortholog(s) (if have; sometimes more than one).<br/>
Each element is a dictionary of <strong>mouse gene name</strong>, <strong>mouse gene id</strong>, <strong>homology type</strong> (one to one/one to multiple/many to many), and <strong>similarity of human and mouse gene in percentage</strong>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">h2m</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">'U1'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    Query human gene: U1;
    Mouse ortholog(s): Gm22866,Gm25938;
    Homology type: one2many;
    Sequence Simalarity(%):68.75, 62.3457.
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">h2m</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">'TPT1P6'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    The query human gene: TPT1P6 has no mouse ortholog or this gene id is not included in the database. Please check the input format.
</pre></div>
</div>
<p>Except for gene names, both ENSEMBL gene id and transcript id are accepted to identify a human gene. You can use the <strong>ty</strong> parameter (‘tx_id’,’gene_id’ or ‘name’) to specify your input type, but this is totally optional.</p>
<p>Using gene id:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">query_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">'ENSG00000141510'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    Query human gene: TP53;
    Mouse ortholog(s): Trp53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.
</pre></div>
</div>
<p>Using transcript id. Should include a db annotation file with the same ref genome version.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">query_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">'ENST00000269305.4'</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="n">db_h</span><span class="p">,</span> <span class="n">ty</span><span class="o">=</span><span class="s1">'tx_id'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    Query human gene: TP53;
    Mouse ortholog(s): Trp53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.
</pre></div>
</div>
<p>The query result of all human genes, as well as corresponding transcript IDs, is also available as a csv file in the <a href="https://www.dropbox.com/scl/fo/1wtrnc9w6s9gemweuw2fv/h?rlkey=hli1z6tv096cjwit5oi6bwggg&amp;dl=0" rel="noopener noreferrer" target="_blank">this Dropbox folder</a>.</p>
</section>
<section id="get-transcript-id">
<h3>Get transcript ID<a class="headerlink" href="#get-transcript-id" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#get-transcript-id'"><span>#</span></a></h3>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Internet connection needed for this function</p>
</div>
<p>One gene may have different transcripts. For mutation modeling, it is important to specify one transcript. If you do not have this information in hand, you can use H2M to get it.</p>
<p>Again, both gene IDs and gene names are accepted as identificaitons for human and mouse genes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">list_tx_id_h</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_id</span><span class="p">(</span><span class="s1">'TP53'</span><span class="p">,</span> <span class="s1">'h'</span><span class="p">,</span> <span class="n">ver</span><span class="o">=</span><span class="mi">37</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>  Genome assembly: GRCh37;
  The canonical transcript is: ENST00000269305.4;
  You can choose from the 17 transcripts below for further analysis:
  (1)ENST00000269305.4 (2)ENST00000413465.2 (3)ENST00000359597.4 (4)ENST00000504290.1 (5)ENST00000510385.1 (6)ENST00000504937.1 (7)ENST00000455263.2 (8)ENST00000420246.2 (9)ENST00000445888.2 (10)ENST00000576024.1 (11)ENST00000509690.1 (12)ENST00000514944.1 (13)ENST00000574684.1 (14)ENST00000505014.1 (15)ENST00000508793.1 (16)ENST00000604348.1 (17)ENST00000503591.1
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">list_tx_id_m</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_id</span><span class="p">(</span><span class="s1">'ENSMUSG00000059552'</span><span class="p">,</span> <span class="s1">'m'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>  Genome assembly: GRCm39;
  The canonical transcript is: ENSMUST00000108658.10;
  You can choose from the 6 transcripts below for further analysis:
  (1)ENSMUST00000108658.10 (2)ENSMUST00000171247.8 (3)ENSMUST00000005371.12 (4)ENSMUST00000147512.2 (5)ENSMUST00000108657.4 (6)ENSMUST00000130540.2
</pre></div>
</div>
<p>Now you can use H2M to model your human mutations of interest.<br/>
You should have at least such information in hand:</p>
<ol class="arabic simple">
<li><p>Transcript ID of the human gene</p></li>
<li><p>Transcript ID of the mouse gene</p></li>
</ol>
<p>Also, multiple infomation for the huaman variant in <strong>MAF</strong> format:</p>
<ol class="arabic simple" start="3">
<li><p>Start postion</p></li>
<li><p>End position</p></li>
<li><p>Reference sequence</p></li>
<li><p>Alternate sequence</p></li>
<li><p>Type in <code class="docutils literal notranslate"><span class="pre">'SNP','DNP','TNP','ONP','INS','DEL'</span></code></p></li>
<li><p>The version number of human ref genome <code class="docutils literal notranslate"><span class="pre">'37','38'</span></code></p></li>
</ol>
</section>
<section id="modeling-human-variants-in-the-mouse-genome">
<h3>Modeling human variants in the mouse genome<a class="headerlink" href="#modeling-human-variants-in-the-mouse-genome" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#modeling-human-variants-in-the-mouse-genome'"><span>#</span></a></h3>
<section id="basic-usage">
<h4>Basic usage<a class="headerlink" href="#basic-usage" title="Link to this heading"><span>#</span></a></h4>
<p>Taking <em>TP53</em> R273H (ENST00000269305.4:c.818G&gt;A) as an example.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span> <span class="o">=</span> <span class="n">list_tx_id_h</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">list_tx_id_m</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
    <span class="c1"># use the canonical transcript</span>
    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="s1">'C'</span><span class="p">,</span><span class="s1">'T'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
    <span class="n">model_result</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Key</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>gene_name_h</p></td>
<td><p>TP53</p></td>
</tr>
<tr class="row-odd"><td><p>gene_id_h</p></td>
<td><p>ENSG00000141510.11</p></td>
</tr>
<tr class="row-even"><td><p>tx_id_h</p></td>
<td><p>ENST00000269305.4</p></td>
</tr>
<tr class="row-odd"><td><p>chr_h</p></td>
<td><p>chr17</p></td>
</tr>
<tr class="row-even"><td><p>exon_num_h</p></td>
<td><p>10</p></td>
</tr>
<tr class="row-odd"><td><p>strand_h</p></td>
<td><p>-</p></td>
</tr>
<tr class="row-even"><td><p>match</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p>start_h</p></td>
<td><p>7577120</p></td>
</tr>
<tr class="row-even"><td><p>end_h</p></td>
<td><p>7577120</p></td>
</tr>
<tr class="row-odd"><td><p>ref_seq_h</p></td>
<td><p>C</p></td>
</tr>
<tr class="row-even"><td><p>alt_seq_h</p></td>
<td><p>T</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSc_h</p></td>
<td><p>ENST00000269305.4:c.818G&gt;A</p></td>
</tr>
<tr class="row-even"><td><p>HGVSp_h</p></td>
<td><p>R273H</p></td>
</tr>
<tr class="row-odd"><td><p>classification_h</p></td>
<td><p>Missense</p></td>
</tr>
<tr class="row-even"><td><p>exon_h</p></td>
<td><p>E_7</p></td>
</tr>
<tr class="row-odd"><td><p>type_h</p></td>
<td><p>SNP</p></td>
</tr>
<tr class="row-even"><td><p>status</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p>class</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>statement</p></td>
<td><p>Class 0: This mutation can be originally modeled.</p></td>
</tr>
<tr class="row-odd"><td><p>flank_size_left</p></td>
<td><p>4aa</p></td>
</tr>
<tr class="row-even"><td><p>flank_size_right</p></td>
<td><p>15aa</p></td>
</tr>
<tr class="row-odd"><td><p>gene_name_m</p></td>
<td><p>Trp53</p></td>
</tr>
<tr class="row-even"><td><p>gene_id_m</p></td>
<td><p>ENSMUSG00000059552.14</p></td>
</tr>
<tr class="row-odd"><td><p>tx_id_m</p></td>
<td><p>ENSMUST00000108658.10</p></td>
</tr>
<tr class="row-even"><td><p>chr_m</p></td>
<td><p>chr11</p></td>
</tr>
<tr class="row-odd"><td><p>exon_num_m</p></td>
<td><p>10</p></td>
</tr>
<tr class="row-even"><td><p>strand_m</p></td>
<td><p>+</p></td>
</tr>
<tr class="row-odd"><td><p>type_m</p></td>
<td><p>SNP</p></td>
</tr>
<tr class="row-even"><td><p>classification_m</p></td>
<td><p>Missense</p></td>
</tr>
<tr class="row-odd"><td><p>exon_m</p></td>
<td><p>E_7</p></td>
</tr>
<tr class="row-even"><td><p>start_m_ori</p></td>
<td><p>69480434</p></td>
</tr>
<tr class="row-odd"><td><p>end_m_ori</p></td>
<td><p>69480434</p></td>
</tr>
<tr class="row-even"><td><p>ref_seq_m_ori</p></td>
<td><p>G</p></td>
</tr>
<tr class="row-odd"><td><p>alt_seq_m_ori</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-even"><td><p>HGVSc_m_ori</p></td>
<td><p>ENSMUST00000108658.10:c.809G&gt;A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m_ori</p></td>
<td><p>R270H</p></td>
</tr>
<tr class="row-even"><td><p>start_m</p></td>
<td><p>69480434</p></td>
</tr>
<tr class="row-odd"><td><p>end_m</p></td>
<td><p>69480434</p></td>
</tr>
<tr class="row-even"><td><p>ref_seq_m</p></td>
<td><p>G</p></td>
</tr>
<tr class="row-odd"><td><p>alt_seq_m</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-even"><td><p>HGVSc_m</p></td>
<td><p>ENSMUST00000108658.10:c.809G&gt;A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m</p></td>
<td><p>R270H</p></td>
</tr>
</tbody>
</table>
<p>We can see that this human mutaton can be originally modeled by introducing the same neucleotide alteration.</p>
<section id="flank-size">
<h5>Flank Size<a class="headerlink" href="#flank-size" title="Link to this heading"><span>#</span></a></h5>
<p>The length of the identical sequences between human and mouse on teh left/right side of the mutation is provided in order to give you a sense of the local homology and how confident you should be in the fidelity of this modeling.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'flank_size_left'</span><span class="p">,</span><span class="s1">'flank_size_right'</span><span class="p">]]</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>flank_size_left</p></th>
<th class="head"><p>flank_size_left</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>4aa</p></td>
<td><p>15aa</p></td>
</tr>
</tbody>
</table>
</section>
<section id="result-visualization">
<h5>Result visualization<a class="headerlink" href="#result-visualization" title="Link to this heading"><span>#</span></a></h5>
<p>By setting <code class="docutils literal notranslate"><span class="pre">show_sequence</span> <span class="pre">=</span> <span class="pre">True</span></code>, we can output the sequences of the wild-type and mutated human gene, wild-type, originally-modeled, and alternatively-modeled (if exsist) mouse gene. Modeling results with <code class="docutils literal notranslate"><span class="pre">show_sequence</span> <span class="pre">=</span> <span class="pre">True</span></code> can be directly visulaized by <code class="docutils literal notranslate"><span class="pre">h2m.visulization</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="s1">'C'</span><span class="p">,</span><span class="s1">'T'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">,</span> <span class="n">show_sequence</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)</span>
    <span class="n">h2m</span><span class="o">.</span><span class="n">visualization</span><span class="p">(</span><span class="n">model_result</span><span class="p">,</span> <span class="n">flank_size</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">print_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/h2m_visual.png"><img alt="_images/h2m_visual.png" class="align-center" src="_images/h2m_visual.png" style="width: 800px;"/></a>
</section>
</section>
<section id="alternative-modeling">
<h4>Alternative modeling<a class="headerlink" href="#alternative-modeling" title="Link to this heading"><span>#</span></a></h4>
<p>Sometimes the human mutation cannot be originally modeled in the mouse genome by using the same neucleotide alteration. Under this circumsatance, some alternative modeling strategies may be found by searching the codon list of the target amino acids.</p>
<ul class="simple">
<li><p>Example 1: TP53 R306Q.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577021</span><span class="p">,</span> <span class="mi">7577021</span><span class="p">,</span> <span class="s1">'C'</span><span class="p">,</span><span class="s1">'T'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'HGVSc_h'</span><span class="p">,</span><span class="s1">'HGVSp_h'</span><span class="p">,</span>
                                <span class="s1">'HGVSc_m_ori'</span><span class="p">,</span><span class="s1">'HGVSp_m_ori'</span><span class="p">,</span>
                                <span class="s1">'HGVSc_m'</span><span class="p">,</span><span class="s1">'HGVSp_m'</span><span class="p">]]</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p></p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>HGVSc_h</p></td>
<td><p>ENST00000269305.4:c.917G&gt;A</p></td>
<td><p>ENST00000269305.4:c.917G&gt;A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_h</p></td>
<td><p>R306Q</p></td>
<td><p>R306Q</p></td>
</tr>
<tr class="row-even"><td><p>HGVSc_m_ori</p></td>
<td><p>ENSMUST00000108658.10:c.908G&gt;A</p></td>
<td><p>ENSMUST00000108658.10:c.908G&gt;A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m_ori</p></td>
<td><p>R303K</p></td>
<td><p>R303K</p></td>
</tr>
<tr class="row-even"><td><p>HGVSc_m</p></td>
<td><p>ENSMUST00000108658.10:c.907_908AG&gt;CA</p></td>
<td><p>ENSMUST00000108658.10:c.907_909AGA&gt;CAG</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m</p></td>
<td><p>R303Q</p></td>
<td><p>R303Q</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p>Example 2: TP53 R249_T253delinsS.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577523</span><span class="p">,</span> <span class="mi">7577534</span><span class="p">,</span> <span class="s1">'GTGAGGATGGGC'</span><span class="p">,</span> <span class="s1">'-'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'DEL'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'HGVSc_h'</span><span class="p">,</span><span class="s1">'HGVSp_h'</span><span class="p">,</span>
                                <span class="s1">'HGVSc_m_ori'</span><span class="p">,</span><span class="s1">'HGVSp_m_ori'</span><span class="p">,</span>
                                <span class="s1">'HGVSc_m'</span><span class="p">,</span><span class="s1">'HGVSp_m'</span><span class="p">]]</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/delins.png"><img alt="_images/delins.png" class="align-center" src="_images/delins.png" style="width: 800px;"/></a>
<p>The default maximum number of output alternatives is 5. You can definitly change that by the parameter <strong>max_alternative</strong>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">model_result_long</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577523</span><span class="p">,</span> <span class="mi">7577534</span><span class="p">,</span> <span class="s1">'GTGAGGATGGGC'</span><span class="p">,</span> <span class="s1">'-'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'DEL'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">,</span> <span class="n">max_alternative</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">model_result</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">model_result_long</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>(5, 6)
</pre></div>
</div>
<p>If you do not want to alternatively model variants, you can set <strong>search_alternatve</strong> to False.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577523</span><span class="p">,</span> <span class="mi">7577534</span><span class="p">,</span> <span class="s1">'GTGAGGATGGGC'</span><span class="p">,</span> <span class="s1">'-'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'DEL'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">,</span> <span class="n">search_alternative</span><span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">model_result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">'statement'</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>'Class 6: This mutation cannot be originally modeled.'
</pre></div>
</div>
</section>
<section id="original-modeling-with-uncertain-effects">
<h4>Original modeling with uncertain effects<a class="headerlink" href="#original-modeling-with-uncertain-effects" title="Link to this heading"><span>#</span></a></h4>
<p>For frame-shifting mutations and mutations in the non-coding region, we cannot find such alternative modeling strategies with the same protein change effects. H2M will only offer the original modeling and its effect.</p>
<ul class="simple">
<li><p>Example 1: <em>TP53</em> C275Lfs*31</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> 
                                <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> 
                                <span class="mi">7577115</span><span class="p">,</span> <span class="mi">7577116</span><span class="p">,</span> <span class="s1">''</span><span class="p">,</span><span class="s1">'A'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'INS'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'HGVSc_h'</span><span class="p">,</span><span class="s1">'HGVSp_h'</span><span class="p">,</span>
                                <span class="s1">'HGVSc_m'</span><span class="p">,</span><span class="s1">'HGVSp_m'</span><span class="p">]]</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>HGVSc_h</p></td>
<td><p>ENST00000269305.4:c.822_823&gt;T</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_h</p></td>
<td><p>C275Lfs*31</p></td>
</tr>
<tr class="row-even"><td><p>HGVSc_m</p></td>
<td><p>ENSMUST00000108658.10:c.813_814&gt;T</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m</p></td>
<td><p>C272Lfs*24</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p>Example 2: TP53 splice site mutation</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> 
                            <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7578555</span><span class="p">,</span> <span class="mi">7578555</span><span class="p">,</span> 
                            <span class="s1">'C'</span><span class="p">,</span> <span class="s1">'T'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'HGVSc_h'</span><span class="p">,</span><span class="s1">'HGVSp_h'</span><span class="p">,</span>
                            <span class="s1">'HGVSc_m'</span><span class="p">,</span><span class="s1">'HGVSp_m'</span><span class="p">]]</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>HGVSc_h</p></td>
<td><p>ENST00000269305.4:c.376-1G&gt;A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_h</p></td>
<td><p>X125_splice</p></td>
</tr>
<tr class="row-even"><td><p>HGVSc_m</p></td>
<td><p>ENSMUST00000108658.10:c.367-1G&gt;A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m</p></td>
<td><p>X122_splice</p></td>
</tr>
</tbody>
</table>
</section>
</section>
</section>
<section id="additional-usage-hint">
<h2>Additional Usage Hint<a class="headerlink" href="#additional-usage-hint" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#additional-usage-hint'"><span>#</span></a></h2>
<section id="additional-function-1-modeling-m2h">
<h3>Additional function 1: modeling M2H<a class="headerlink" href="#additional-function-1-modeling-m2h" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#additional-function-1-modeling-m2h'"><span>#</span></a></h3>
<p>Replace human variant coordinates and sequences with murine ones, and set <code class="docutils literal notranslate"><span class="pre">direction</span> <span class="pre">=</span> <span class="pre">'m2h'</span></code>.  Use TP53 R273H as an example.</p>
<p>H2M:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="mi">7577120</span><span class="p">,</span> <span class="s1">'C'</span><span class="p">,</span><span class="s1">'T'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'start_h'</span><span class="p">,</span><span class="s1">'end_h'</span><span class="p">,</span><span class="s1">'ref_seq_h'</span><span class="p">,</span><span class="s1">'alt_seq_h'</span><span class="p">,</span><span class="s1">'HGVSp_h'</span><span class="p">,</span><span class="s1">'start_m'</span><span class="p">,</span><span class="s1">'end_m'</span><span class="p">,</span><span class="s1">'ref_seq_m'</span><span class="p">,</span><span class="s1">'alt_seq_m'</span><span class="p">,</span><span class="s1">'HGVSp_m'</span><span class="p">]]</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>start_h</p></td>
<td><p>7577120</p></td>
</tr>
<tr class="row-odd"><td><p>end_h</p></td>
<td><p>7577120</p></td>
</tr>
<tr class="row-even"><td><p>ref_seq_h</p></td>
<td><p>C</p></td>
</tr>
<tr class="row-odd"><td><p>alt_seq_h</p></td>
<td><p>T</p></td>
</tr>
<tr class="row-even"><td><p>HGVSp_h</p></td>
<td><p>R273H</p></td>
</tr>
<tr class="row-odd"><td><p>start_m</p></td>
<td><p>69480434</p></td>
</tr>
<tr class="row-even"><td><p>end_m</p></td>
<td><p>69480434</p></td>
</tr>
<tr class="row-odd"><td><p>ref_seq_m</p></td>
<td><p>G</p></td>
</tr>
<tr class="row-even"><td><p>alt_seq_m</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m</p></td>
<td><p>R270H</p></td>
</tr>
</tbody>
</table>
<p>M2H:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> 
                            <span class="mi">69480434</span><span class="p">,</span> <span class="mi">69480434</span><span class="p">,</span> <span class="s1">'G'</span><span class="p">,</span> <span class="s1">'A'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">,</span> 
                            <span class="n">direction</span><span class="o">=</span><span class="s1">'m2h'</span><span class="p">)</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'start_h'</span><span class="p">,</span><span class="s1">'end_h'</span><span class="p">,</span><span class="s1">'ref_seq_h'</span><span class="p">,</span><span class="s1">'alt_seq_h'</span><span class="p">,</span><span class="s1">'HGVSp_h'</span><span class="p">,</span><span class="s1">'start_m'</span><span class="p">,</span><span class="s1">'end_m'</span><span class="p">,</span><span class="s1">'ref_seq_m'</span><span class="p">,</span><span class="s1">'alt_seq_m'</span><span class="p">,</span><span class="s1">'HGVSp_m'</span><span class="p">]]</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>start_h</p></td>
<td><p>7577120</p></td>
</tr>
<tr class="row-odd"><td><p>end_h</p></td>
<td><p>7577120</p></td>
</tr>
<tr class="row-even"><td><p>ref_seq_h</p></td>
<td><p>C</p></td>
</tr>
<tr class="row-odd"><td><p>alt_seq_h</p></td>
<td><p>T</p></td>
</tr>
<tr class="row-even"><td><p>HGVSp_h</p></td>
<td><p>R273H</p></td>
</tr>
<tr class="row-odd"><td><p>start_m</p></td>
<td><p>69480434</p></td>
</tr>
<tr class="row-even"><td><p>end_m</p></td>
<td><p>69480434</p></td>
</tr>
<tr class="row-odd"><td><p>ref_seq_m</p></td>
<td><p>G</p></td>
</tr>
<tr class="row-even"><td><p>alt_seq_m</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m</p></td>
<td><p>R270H</p></td>
</tr>
</tbody>
</table>
</section>
<section id="additional-function-2-modeling-h2h-m2m-paralogs">
<h3>Additional function 2: modeling H2H/M2M paralogs<a class="headerlink" href="#additional-function-2-modeling-h2h-m2m-paralogs" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#additional-function-2-modeling-h2h-m2m-paralogs'"><span>#</span></a></h3>
<p>Replace the reference genome and gencode annotation database input parameter to do so.  Take human IDH1 R172G as an example.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">'class'</span><span class="p">]</span><span class="o">==</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tx_id_1_h</span><span class="p">,</span> <span class="n">tx_id_2_h</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_id</span><span class="p">(</span><span class="s1">'SMARCA2'</span><span class="p">,</span><span class="s1">'h'</span><span class="p">,</span><span class="n">ver</span><span class="o">=</span><span class="mi">37</span><span class="p">)[</span><span class="mi">3</span><span class="p">],</span><span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_id</span><span class="p">(</span><span class="s1">'SMARCA4'</span><span class="p">,</span><span class="s1">'h'</span><span class="p">,</span><span class="n">ver</span><span class="o">=</span><span class="mi">37</span><span class="p">)[</span><span class="mi">3</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Genome assembly: GRCh37;
The canonical transcript is: ENST00000382203.1;
You can choose from the 17 transcripts below for further analysis:
(1)ENST00000382203.1 (2)ENST00000450198.1 (3)ENST00000457226.1 (4)ENST00000439732.1 (5)ENST00000382194.1 (6)ENST00000491574.1 (7)ENST00000452193.1 (8)ENST00000302401.3 (9)ENST00000423555.1 (10)ENST00000382186.1 (11)ENST00000417599.1 (12)ENST00000382185.1 (13)ENST00000382183.1 (14)ENST00000416751.1 (15)ENST00000349721.2 (16)ENST00000357248.2 (17)ENST00000324954.5

Genome assembly: GRCh37;
The canonical transcript is: ENST00000429416.3;
You can choose from the 20 transcripts below for further analysis:
(1)ENST00000429416.3 (2)ENST00000344626.4 (3)ENST00000541122.2 (4)ENST00000589677.1 (5)ENST00000444061.3 (6)ENST00000590574.1 (7)ENST00000591545.1 (8)ENST00000592604.1 (9)ENST00000586122.1 (10)ENST00000587988.1 (11)ENST00000591595.1 (12)ENST00000585799.1 (13)ENST00000592158.1 (14)ENST00000586892.1 (15)ENST00000538456.3 (16)ENST00000586985.1 (17)ENST00000586921.1 (18)ENST00000358026.2 (19)ENST00000413806.3 (20)ENST00000450717.3
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_h</span><span class="p">,</span> <span class="n">index_list_h</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">tx_id_1_h</span><span class="p">,</span> <span class="n">tx_id_2_h</span><span class="p">,</span> 
                        <span class="mi">2115855</span><span class="p">,</span> <span class="mi">2115855</span><span class="p">,</span> <span class="s1">'G'</span><span class="p">,</span> <span class="s1">'A'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">,</span>
                        <span class="n">direction</span><span class="o">=</span><span class="s1">'h2h'</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'gene_name_h_1'</span><span class="p">,</span><span class="s1">'start_h_1'</span><span class="p">,</span><span class="s1">'end_h_1'</span><span class="p">,</span><span class="s1">'ref_seq_h_1'</span><span class="p">,</span><span class="s1">'alt_seq_h_1'</span><span class="p">,</span><span class="s1">'HGVSp_h_1'</span><span class="p">,</span><span class="s1">'gene_name_h_2'</span><span class="p">,</span><span class="s1">'start_h_2'</span><span class="p">,</span><span class="s1">'end_h_2'</span><span class="p">,</span><span class="s1">'ref_seq_h_2'</span><span class="p">,</span><span class="s1">'alt_seq_h_2'</span><span class="p">,</span><span class="s1">'HGVSp_h_2'</span><span class="p">]]</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>gene_name_h_1</p></td>
<td><p>SMARCA2</p></td>
</tr>
<tr class="row-odd"><td><p>start_h_1</p></td>
<td><p>2115855</p></td>
</tr>
<tr class="row-even"><td><p>end_h_1</p></td>
<td><p>2115855</p></td>
</tr>
<tr class="row-odd"><td><p>ref_seq_h_1</p></td>
<td><p>G</p></td>
</tr>
<tr class="row-even"><td><p>alt_seq_h_1</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_h_1</p></td>
<td><p>G1164R</p></td>
</tr>
<tr class="row-even"><td><p>gene_name_h_2</p></td>
<td><p>SMARCA4</p></td>
</tr>
<tr class="row-odd"><td><p>start_h_2</p></td>
<td><p>11143999</p></td>
</tr>
<tr class="row-even"><td><p>end_h_2</p></td>
<td><p>11143999</p></td>
</tr>
<tr class="row-odd"><td><p>ref_seq_h_2</p></td>
<td><p>G</p></td>
</tr>
<tr class="row-even"><td><p>alt_seq_h_2</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_h_2</p></td>
<td><p>G1194R</p></td>
</tr>
</tbody>
</table>
</section>
<section id="additional-function-3-modeling-for-base-editing">
<h3>Additional function 3: modeling for base editing<a class="headerlink" href="#additional-function-3-modeling-for-base-editing" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#additional-function-3-modeling-for-base-editing'"><span>#</span></a></h3>
<p>When you set <strong>param = ‘BE’</strong>, you will get modeling results that can be modeled by base editing (A-&gt;G, G-&gt;A, C-&gt;T, T-&gt;C, AA-&gt;GG, …etc.). If one mutation can be originally modeled in the mouse genome but not in a BE style, alternative BE modeling strategies will be returned too.</p>
<p>Taking <em>KEAP1</em> F221L as an example.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">h2m</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">'KEAP1'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    Query human gene: KEAP1;
    Mouse ortholog(s): Keap1;
    Homology type: one2one;
    Sequence Simalarity(%):94.0705.
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">tx_id_h_2</span><span class="p">,</span> <span class="n">tx_id_m_2</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_id</span><span class="p">(</span><span class="s1">'KEAP1'</span><span class="p">,</span><span class="s1">'h'</span><span class="p">,</span><span class="n">ver</span><span class="o">=</span><span class="mi">37</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="mi">3</span><span class="p">],</span> <span class="n">h2m</span><span class="o">.</span><span class="n">get_tx_id</span><span class="p">(</span><span class="s1">'Keap1'</span><span class="p">,</span><span class="s1">'m'</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="mi">3</span><span class="p">]</span>
    <span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h_2</span><span class="p">,</span> <span class="n">tx_id_m_2</span><span class="p">,</span> <span class="mi">10602915</span><span class="p">,</span> <span class="mi">10602915</span><span class="p">,</span> <span class="s1">'G'</span><span class="p">,</span><span class="s1">'T'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">,</span> <span class="n">param</span><span class="o">=</span><span class="s1">'BE'</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)[[</span><span class="s1">'HGVSc_h'</span><span class="p">,</span><span class="s1">'HGVSp_h'</span><span class="p">,</span><span class="s1">'HGVSc_m_ori'</span><span class="p">,</span><span class="s1">'HGVSp_m_ori'</span><span class="p">,</span><span class="s1">'statement'</span><span class="p">,</span><span class="s1">'HGVSc_m'</span><span class="p">,</span><span class="s1">'HGVSp_m'</span><span class="p">]]</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>HGVSc_h</p></td>
<td><p>ENST00000171111.5:c.663C&gt;A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_h</p></td>
<td><p>F221L</p></td>
</tr>
<tr class="row-even"><td><p>HGVSc_m_ori</p></td>
<td><p>ENSMUST00000164812.8:c.663C&gt;A</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSp_m_ori</p></td>
<td><p>F221L</p></td>
</tr>
<tr class="row-even"><td><p>statement</p></td>
<td><p>Class 1: This mutation can be alternatively modeled.</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSc_m</p></td>
<td><p>ENSMUST00000164812.8:c.661T&gt;C</p></td>
</tr>
<tr class="row-even"><td><p>HGVSp_m</p></td>
<td><p>F221L</p></td>
</tr>
</tbody>
</table>
</section>
<section id="additional-function-4-modeling-by-amino-acid-change-input">
<h3>Additional function 4: modeling by amino acid change input<a class="headerlink" href="#additional-function-4-modeling-by-amino-acid-change-input" title="Link to this heading" x-intersect.margin.0%.0%.-70%.0%="activeSection = '#additional-function-4-modeling-by-amino-acid-change-input'"><span>#</span></a></h3>
<p>Set <strong>coor = ‘aa’</strong> and modeling variants by amino acid change input. Use TP53 R175H as an example.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">model_result</span> <span class="o">=</span> <span class="n">h2m</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">records_h</span><span class="p">,</span><span class="n">index_list_h</span><span class="p">,</span> <span class="n">records_m</span><span class="p">,</span> <span class="n">index_list_m</span><span class="p">,</span> <span class="n">db_h</span><span class="p">,</span> <span class="n">db_m</span><span class="p">,</span> <span class="n">tx_id_h</span><span class="p">,</span> <span class="n">tx_id_m</span><span class="p">,</span> <span class="mi">175</span><span class="p">,</span> <span class="mi">175</span><span class="p">,</span> <span class="s1">'R'</span><span class="p">,</span> <span class="s1">'H'</span><span class="p">,</span> <span class="n">coor</span> <span class="o">=</span> <span class="s1">'aa'</span><span class="p">,</span> <span class="n">ty_h</span> <span class="o">=</span> <span class="s1">'SNP'</span><span class="p">,</span> <span class="n">ver</span> <span class="o">=</span> <span class="mi">37</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model_result</span><span class="p">)</span>
</pre></div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Key</p></th>
<th class="head"><p>0</p></th>
<th class="head"><p>1</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>gene_name_h</p></td>
<td><p>TP53</p></td>
<td><p>TP53</p></td>
</tr>
<tr class="row-odd"><td><p>gene_id_h</p></td>
<td><p>ENSG00000141510.11</p></td>
<td><p>ENSG00000141510.11</p></td>
</tr>
<tr class="row-even"><td><p>tx_id_h</p></td>
<td><p>ENST00000269305.4</p></td>
<td><p>ENST00000269305.4</p></td>
</tr>
<tr class="row-odd"><td><p>chr_h</p></td>
<td><p>chr17</p></td>
<td><p>chr17</p></td>
</tr>
<tr class="row-even"><td><p>exon_num_h</p></td>
<td><p>10</p></td>
<td><p>10</p></td>
</tr>
<tr class="row-odd"><td><p>strand_h</p></td>
<td><p>-</p></td>
<td><p>-</p></td>
</tr>
<tr class="row-even"><td><p>match</p></td>
<td><p>True</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p>start_h</p></td>
<td><p>7578405</p></td>
<td><p>7578405</p></td>
</tr>
<tr class="row-even"><td><p>end_h</p></td>
<td><p>7578407</p></td>
<td><p>7578407</p></td>
</tr>
<tr class="row-odd"><td><p>ref_seq_m_ori</p></td>
<td><p>CGC</p></td>
<td><p>CGC</p></td>
</tr>
<tr class="row-even"><td><p>alt_seq_m_ori</p></td>
<td><p>CAC</p></td>
<td><p>CAC</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSc_m_ori</p></td>
<td><p>ENSMUST00000108658.10:c.514_516CGC&gt;CAC</p></td>
<td><p>ENSMUST00000108658.10:c.514_516CGC&gt;CAC</p></td>
</tr>
<tr class="row-even"><td><p>HGVSp_m_ori</p></td>
<td><p>R172H</p></td>
<td><p>R172H</p></td>
</tr>
<tr class="row-odd"><td><p>start_m</p></td>
<td><p>69479338</p></td>
<td><p>69479338</p></td>
</tr>
<tr class="row-even"><td><p>end_m</p></td>
<td><p>69479338</p></td>
<td><p>69479339</p></td>
</tr>
<tr class="row-odd"><td><p>ref_seq_m</p></td>
<td><p>G</p></td>
<td><p>GC</p></td>
</tr>
<tr class="row-even"><td><p>alt_seq_m</p></td>
<td><p>A</p></td>
<td><p>AT</p></td>
</tr>
<tr class="row-odd"><td><p>HGVSc_m</p></td>
<td><p>ENSMUST00000108658.10:c.515G&gt;A</p></td>
<td><p>ENSMUST00000108658.10:c.515_516GC&gt;AT</p></td>
</tr>
<tr class="row-even"><td><p>HGVSp_m</p></td>
<td><p>R172H</p></td>
<td><p>R172H</p></td>
</tr>
</tbody>
</table>
<p>All of these can also be done in a batch-processing style by using <code class="docutils literal notranslate"><span class="pre">h2m.model_batch</span></code>.</p>
</section>
</section>
</section>
</div></div><aside class="hidden text-sm xl:block" id="right-sidebar">
<div class="sticky top-16 -mt-10 max-h-[calc(100vh-5rem)] overflow-y-auto pt-6 space-y-2"><p class="font-medium">On this page</p>
<ul>
<li><a :data-current="activeSection === '#package-installation'" class="reference internal" href="#package-installation">Package Installation</a></li>
<li><a :data-current="activeSection === '#importing-packages'" class="reference internal" href="#importing-packages">Importing packages</a></li>
<li><a :data-current="activeSection === '#loading-data'" class="reference internal" href="#loading-data">Loading data</a></li>
<li><a :data-current="activeSection === '#batch-processing'" class="reference internal" href="#batch-processing">Batch Processing</a><ul>
<li><a :data-current="activeSection === '#input-format'" class="reference internal" href="#input-format">Input format</a><ul>
<li><a :data-current="activeSection === '#read-from-cbioportal-maf'" class="reference internal" href="#read-from-cbioportal-maf">Read from cBioPortal - MAF</a></li>
<li><a :data-current="activeSection === '#read-from-genomad-vcf'" class="reference internal" href="#read-from-genomad-vcf">Read from GenomAD  - VCF</a></li>
<li><a :data-current="activeSection === '#read-from-clinvar'" class="reference internal" href="#read-from-clinvar">Read from ClinVar</a></li>
</ul>
</li>
<li><a :data-current="activeSection === '#get-canonical-transcript-ids-for-human'" class="reference internal" href="#get-canonical-transcript-ids-for-human">Get canonical transcript IDs for human</a></li>
<li><a :data-current="activeSection === '#query-orthologous-genes'" class="reference internal" href="#query-orthologous-genes">Query orthologous genes</a></li>
<li><a :data-current="activeSection === '#get-canonical-transcript-ids-for-mouse'" class="reference internal" href="#get-canonical-transcript-ids-for-mouse">Get canonical transcript IDs for mouse</a></li>
<li><a :data-current="activeSection === '#compute-the-muerine-variant-equivalents'" class="reference internal" href="#compute-the-muerine-variant-equivalents">Compute the muerine variant equivalents</a></li>
</ul>
</li>
<li><a :data-current="activeSection === '#single-variant-input'" class="reference internal" href="#single-variant-input">Single variant input</a><ul>
<li><a :data-current="activeSection === '#id1'" class="reference internal" href="#id1">Query orthologous genes</a></li>
<li><a :data-current="activeSection === '#get-transcript-id'" class="reference internal" href="#get-transcript-id">Get transcript ID</a></li>
<li><a :data-current="activeSection === '#modeling-human-variants-in-the-mouse-genome'" class="reference internal" href="#modeling-human-variants-in-the-mouse-genome">Modeling human variants in the mouse genome</a><ul>
<li><a :data-current="activeSection === '#basic-usage'" class="reference internal" href="#basic-usage">Basic usage</a><ul>
<li><a :data-current="activeSection === '#flank-size'" class="reference internal" href="#flank-size">Flank Size</a></li>
<li><a :data-current="activeSection === '#result-visualization'" class="reference internal" href="#result-visualization">Result visualization</a></li>
</ul>
</li>
<li><a :data-current="activeSection === '#alternative-modeling'" class="reference internal" href="#alternative-modeling">Alternative modeling</a></li>
<li><a :data-current="activeSection === '#original-modeling-with-uncertain-effects'" class="reference internal" href="#original-modeling-with-uncertain-effects">Original modeling with uncertain effects</a></li>
</ul>
</li>
</ul>
</li>
<li><a :data-current="activeSection === '#additional-usage-hint'" class="reference internal" href="#additional-usage-hint">Additional Usage Hint</a><ul>
<li><a :data-current="activeSection === '#additional-function-1-modeling-m2h'" class="reference internal" href="#additional-function-1-modeling-m2h">Additional function 1: modeling M2H</a></li>
<li><a :data-current="activeSection === '#additional-function-2-modeling-h2h-m2m-paralogs'" class="reference internal" href="#additional-function-2-modeling-h2h-m2m-paralogs">Additional function 2: modeling H2H/M2M paralogs</a></li>
<li><a :data-current="activeSection === '#additional-function-3-modeling-for-base-editing'" class="reference internal" href="#additional-function-3-modeling-for-base-editing">Additional function 3: modeling for base editing</a></li>
<li><a :data-current="activeSection === '#additional-function-4-modeling-by-amino-acid-change-input'" class="reference internal" href="#additional-function-4-modeling-by-amino-acid-change-input">Additional function 4: modeling by amino acid change input</a></li>
</ul>
</li>
</ul>
</div>
</aside>
</main>
</div>
</div><footer class="py-6 border-t border-border md:py-0">
<div class="container flex flex-col items-center justify-between gap-4 md:h-24 md:flex-row">
<div class="flex flex-col items-center gap-4 px-8 md:flex-row md:gap-2 md:px-0">
<p class="text-sm leading-loose text-center text-muted-foreground md:text-left">© 2024, Kexin Dong Built with <a class="font-medium underline underline-offset-4" href="https://www.sphinx-doc.org" rel="noreferrer">Sphinx 7.2.6</a></p>
</div>
</div>
</footer>
</div>
<script src="_static/documentation_options.js?v=8d563738"></script>
<script src="_static/doctools.js?v=888ff710"></script>
<script src="_static/sphinx_highlight.js?v=dc90522c"></script>
<script defer="defer" src="_static/theme.js?v=8dc3ddfe"></script>
</body>
</html>